storage/src/storage_backend.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) SimpleStaking and Tezedge Contributors
+// SPDX-License-Identifier: MIT
+
+use crate::persistent::database::{DBError, RocksDBStats};
+use failure::Fail;
+use serde::Serialize;
+use std::collections::HashSet;
+use std::mem;
+
+use crate::merkle_storage::{ContextValue, EntryHash};
+
+pub fn size_of_vec<T>(v: &Vec<T>) -> usize {
+    mem::size_of::<Vec<T>>() + mem::size_of::<T>() * v.capacity()
+}
+
+#[derive(Debug, Fail)]
+pub enum StorageBackendError {
+    #[fail(display = "RocksDB error: {}", error)]
+    RocksDBError { error: rocksdb::Error },
+    #[fail(display = "Column family {} is missing", name)]
+    MissingColumnFamily { name: &'static str },
+    #[fail(display = "Backend Error")]
+    BackendError,
+    #[fail(display = "SledDB error: {}", error)]
+    SledDBError { error: sled::Error },
+    #[fail(display = "Guard Poison {} ", error)]
+    GuardPoison { error: String },
+    #[fail(display = "Serialization error: {:?}", error)]
+    SerializationError { error: bincode::Error },
+    #[fail(display = "DBError error: {:?}", error)]
+    DBError { error: DBError },
+}
+
+impl From<rocksdb::Error> for StorageBackendError {
+    fn from(error: rocksdb::Error) -> Self {
+        StorageBackendError::RocksDBError { error }
+    }
+}
+
+impl From<sled::Error> for StorageBackendError {
+    fn from(error: sled::Error) -> Self {
+        StorageBackendError::SledDBError { error }
+    }
+}
+
+impl From<DBError> for StorageBackendError {
+    fn from(error: DBError) -> Self {
+        StorageBackendError::DBError { error }
+    }
+}
+
+impl From<bincode::Error> for StorageBackendError {
+    fn from(error: bincode::Error) -> Self {
+        StorageBackendError::SerializationError { error }
+    }
+}
+
+impl slog::Value for StorageBackendError {
+    fn serialize(
+        &self,
+        _record: &slog::Record,
+        key: slog::Key,
+        serializer: &mut dyn slog::Serializer,
+    ) -> slog::Result {
+        serializer.emit_arguments(key, &format_args!("{}", self))
+    }
+}
+
+pub trait StorageBackend: Send + Sync {
```

Comment:
Dumb question, but what is the difference between this and `KeyValueStoreWithSchema` ? just noticed that the interfaces are very similar.

Reply:
AFAIK original author was not happy with some RocksDB types leaking into abstract interface - not really related to rocksdb.

Reply:
ah right - also StorageBackend comes from GC part where it has additional methods like `start_new_cycle` `retain` etc. So probably we dont wont to infect `KeyValueStoreWithSchema` with those :thinking:

Reply:
Make sense. But it is weird to have both I guess things have to be unified/reworked at some point (maybe adding extra traits in addition to `KeyValueStoreWithSchema`), having this duplication is suspicious.

Reply:
https://simplestakingcom.atlassian.net/browse/TE-432

Reply:
https://simplestakingcom.atlassian.net/browse/TE-432